1_19_R2/src/main/java/me/gamercoder215/mobchip/abstraction/ChipUtil1_19_R2.java

```diff
@@ -0,0 +1,1565 @@
+                EnderCrystal bCrystal = (EnderCrystal) crystal.getBukkitEntity();
+                c.onCrystalDestroyed(bCrystal, fromNMS(s), p == null ? null : Bukkit.getPlayer(p.getUUID()));
+            }
+            public Vec3 getFlyTargetLocation() {
+                Location l = c.getTargetLocation();
+                return new Vec3(l.getX(), l.getY(), l.getZ());
+            }
+            public float getFlySpeed() { return c.getFlyingSpeed(); }
+            public float onHurt(DamageSource s, float damage) { return c.onDamage(fromNMS(s), damage); }
+        };
+    }
+
+    @Override
+    public void setCustomPhase(EnderDragon a, CustomPhase c) {
+        net.minecraft.world.entity.boss.enderdragon.EnderDragon nmsMob = toNMS(a);
+        AbstractDragonPhaseInstance nmsPhase = toNMS(c);
+        try {
+            new EnderDragonPhaseManager(nmsMob).setPhase(nmsPhase.getPhase());
+        } catch (IndexOutOfBoundsException ignored) {}
+    }
+
+    @Override
+    public void knockback(EnderDragon a, List<Entity> list) {
+        net.minecraft.world.entity.boss.enderdragon.EnderDragon nmsMob = toNMS(a);
+
+        try {
+            Method m = net.minecraft.world.entity.boss.enderdragon.EnderDragon.class.getDeclaredMethod("a", List.class);
+            m.setAccessible(true);
+            m.invoke(nmsMob, list.stream().map(ChipUtil1_19_R2::toNMS).collect(Collectors.toList()));
+        } catch (Exception e) {
+            Bukkit.getLogger().severe(e.getClass().getSimpleName());
+            Bukkit.getLogger().severe(e.getMessage());
+            for (StackTraceElement s : e.getStackTrace()) Bukkit.getLogger().severe(s.toString());
+        }
+    }
+
+    @Override
+    public EntityController getController(Mob m) {
+        return new EntityController1_19_R2(m);
+    }
+
+    @Override
+    public EntityNavigation getNavigation(Mob m) {
+        return new EntityNavigation1_19_R2(m);
+    }
+
+    @Override
+    public EntityBody getBody(Mob m) {
+        return new EntityBody1_19_R2(m);
+    }
+
+    public static DamageSource toNMS(EntityDamageEvent.DamageCause c) {
+        return switch (c) {
+            case FIRE -> DamageSource.IN_FIRE;
+            case LIGHTNING -> DamageSource.LIGHTNING_BOLT;
+            case FIRE_TICK -> DamageSource.ON_FIRE;
+            case SUFFOCATION -> DamageSource.IN_WALL;
+            case LAVA -> DamageSource.LAVA;
+            case HOT_FLOOR -> DamageSource.HOT_FLOOR;
+            case CRAMMING -> DamageSource.CRAMMING;
+            case DROWNING -> DamageSource.DROWN;
+            case STARVATION -> DamageSource.STARVE;
+            case CONTACT -> DamageSource.CACTUS;
+            case MAGIC -> DamageSource.MAGIC;
+            case FALL -> DamageSource.FALL;
+            case FLY_INTO_WALL -> DamageSource.FLY_INTO_WALL;
+            case VOID -> DamageSource.OUT_OF_WORLD;
+            case WITHER -> DamageSource.WITHER;
+            //case FALLING_BLOCK -> DamageSource.FALLING_BLOCK;
```

Comment:
Looking at the source code, it seems like that DamageSource requires an entity with the source.
I might change the parser to include the actual event in to input it, remove it entirely, or test it will a null value.